### PR TITLE
Basic support for ParamSpec type checking

### PR DESCRIPTION
**Pull request by JukkaL**
Add support for type checking several ParamSpec use cases (PEP 612).

@hauntsaninja previously added support for semantic analysis of ParamSpec
definitions, and this builds on top that foundation.

The implementation has these main things going on:
 * `ParamSpecType` that is similar to `TypeVarType` but has three "flavors" that
   correspond to `P`, `P.args` and `P.kwargs`
 * `CallableType` represents `Callable[P, T]` if the arguments are 
   (`*args: P.args`, `**kwargs: P.kwargs`) -- and more generally, there can also
   be arbitrary additional prefix arguments
 * Type variables of functions and classes can now be represented using
   `ParamSpecType` in addition to `TypeVarType`

There are still a bunch of TODOs. Some of these are important to address before the
release that includes this. I believe that this is good enough to merge and remaining 
issues can be fixed in follow-up PRs.

Notable missing features include these:
 * `Concatenate`
 * Specifying the value of ParamSpec explicitly (e.g. `Z[[int, str, bool]]`)
 * Various validity checks -- currently only some errors are caught
 * Special case of decorating a method (python/typeshed#6347)

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -976,13 +977,17 @@ def check_func_def(self, defn: FuncItem, typ: CallableType, name: Optional[str])
                                 ctx = typ
                             self.fail(message_registry.FUNCTION_PARAMETER_CANNOT_BE_COVARIANT, ctx)
                     if typ.arg_kinds[i] == nodes.ARG_STAR:
-                        # builtins.tuple[T] is typing.Tuple[T, ...]
-                        arg_type = self.named_generic_type('builtins.tuple',
-                                                           [arg_type])
+                        arg_type = get_proper_type(arg_type)
+                        if not isinstance(arg_type, ParamSpecType):
```
Comment (ilevkivskyi) [inline mypy/checker.py:981]: I think you should update the proper_plugin to not require this. `ParamSpecType` can never by a target of a type alias. (Note `TypeVarType` is already excluded).

File: mypy/expandtype.py [hunk 2]
```diff
@@ -98,7 +99,28 @@ def visit_type_var(self, t: TypeVarType) -> Type:
...
+            return Instance(inst.type, inst.args, line=inst.line,
+                            column=inst.column, erased=True)
+        else:
+            return repl
+
     def visit_callable_type(self, t: CallableType) -> Type:
+        param_spec = t.param_spec()
+        if param_spec is not None:
+            repl = get_proper_type(self.variables.get(param_spec.id))
+            if isinstance(repl, CallableType):
```
Comment (ilevkivskyi) [inline mypy/expandtype.py:116]: TBH it is hard to follow the logic here: could you please add a comment on why do we need to have some non-trivial logic for callable?

File: mypy/typeanal.py [hunk 3]
```diff
@@ -956,20 +983,34 @@ def is_defined_type_var(self, tvar: str, context: Context) -> bool:
...
             self.nesting_level += 1
         try:
-            return t.accept(self)
+            analyzed = t.accept(self)
         finally:
             if nested:
                 self.nesting_level -= 1
+        # Use type(...) to ignore proper/non-proper type distinction.
+        if (not allow_param_spec
+                and type(analyzed) is ParamSpecType
```
Comment (ilevkivskyi) [inline mypy/typeanal.py:1005]: Is this really needed? You can update the plugin instead as I suggested above.

File: mypy/types.py [hunk 4]
```diff
@@ -446,13 +447,63 @@ def deserialize(cls, data: JsonDict) -> 'TypeVarType':
...
+            return f'{n}.kwargs'
+        return n
+
+    def __hash__(self) -> int:
+        return hash((self.id, self.flavor))
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, ParamSpecType):
+            return NotImplemented
+        return self.id == other.id and self.flavor == other.flavor
```
Comment (ilevkivskyi) [inline mypy/types.py:506]: Why `upper_bound` is missing here and in various other places like `is_same_type()`? Add a comment about this.

File: mypy/fixup.py [hunk 5]
```diff
@@ -247,6 +248,9 @@ def visit_type_var(self, tvt: TypeVarType) -> None:
         if tvt.upper_bound is not None:
             tvt.upper_bound.accept(self)
 
+    def visit_param_spec(self, p: ParamSpecType) -> None:
+        pass  # Nothing to descend into.
```
Comment (ilevkivskyi) [inline mypy/fixup.py:252]: What about `.upper_bound`? Also why it is ignored in similar fine-grained related logic?

**Review by sobolevn — COMMENTED** [review]
ParamSpec! 🎉 🎏

File: test-data/unit/check-parameter-specification.test [hunk 6]
```diff
@@ -28,51 +28,303 @@ def foo6(x: Callable[[P], int]) -> None: ...  # E: Invalid location for ParamSpe
...
+
+P = ParamSpec('P')
+
+class C(Generic[P]):
+    def __init__(self, x: Callable[P, None]) -> None: ...
+
+    def m(self, *args: P.args, **kwargs: P.kwargs) -> int:
+        return 1
+
+c: C[Any]
```
Comment (sobolevn) [inline test-data/unit/check-parameter-specification.test:318]: Would any other non-`Any` type arguments work? I haven't see any examples of it in tests 🤔

File: mypy/checker.py (hunk 1 above)
Comment (JukkaL) [reply mypy/checker.py:980]: A good idea! Done. Also removed various now-redundant `get_proper_type` calls.

File: mypy/expandtype.py (hunk 2 above)
Comment (JukkaL) [reply mypy/expandtype.py:123]: Added a comment.

File: mypy/typeanal.py (hunk 3 above)
Comment (JukkaL) [reply mypy/typeanal.py]: Replaced with regular `isinstance`.

File: mypy/types.py (hunk 4 above)
Comment (JukkaL) [reply mypy/types.py:512]: It's not user-configurable and always depends on flavor (actually currently it's always `object`). Added some comments.

File: mypy/fixup.py (hunk 5 above)
Comment (JukkaL) [reply mypy/fixup.py]: I hadn't updated some places after I added `upper_bound`. It's now processed here and fine-grained logic.

File: mypy/type_visitor.py [hunk 7]
```diff
@@ -63,6 +63,11 @@ def visit_deleted_type(self, t: DeletedType) -> T:
     def visit_type_var(self, t: TypeVarType) -> T:
         pass
 
+    @abstractmethod
+    def visit_param_spec(self, t: ParamSpecType) -> T:
+        assert False
```
[the remark this answers is not in the document]
Comment (JukkaL) [reply mypy/type_visitor.py]: I used this for testing. Removed the assert.

File: mypy/types.py [hunk 8]
```diff
@@ -446,13 +447,63 @@ def deserialize(cls, data: JsonDict) -> 'TypeVarType':
         )
 
 
+class ParamSpecFlavor:
```
[the remark this answers is not in the document]
Comment (JukkaL) [reply mypy/types.py:450]: I tend to avoid enums since mypyc only partially supports them, so it's possible to accidentally write code that it's pretty slow.

File: test-data/unit/check-parameter-specification.test (hunk 6 above)
Comment (JukkaL) [reply test-data/unit/check-parameter-specification.test:318]: This is not implemented. I've mentioned it in the commit summary as follow-up work.

**Comment by JukkaL** [comment]
Thanks for the feedback! I think that I've addressed all the comments (either by updating the PR or by leaving a reply).

**Review by ilevkivskyi — APPROVED** [review]

**Comment by JukkaL** [comment]
This is still broken:
```py
import atexit

def f() -> None: pass
atexit.register(lambda: f())  # Cannot infer type of lambda
```

I'll fix this in a follow-up PR.
